### PR TITLE
New keyboard: Model H controller for IBM Model M

### PR DESCRIPTION
**Pull request by jberclaz**
## Description

New keyboard: Model H controller for IBM Model M

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/modelh/rules.mk [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# MCU name
+MCU = STM32F103
+
+# GENERIC STM32F103C8T6 board - stm32duino bootloader
+OPT_DEFS = -DCORTEX_VTOR_INIT=0x2000
+MCU_LDSCRIPT = STM32F103x8_stm32duino_bootloader
+BOARD = STM32_F103_STM32DUINO
+
+DFU_ARGS = -d 1eaf:0003 -a2 -R
+DFU_SUFFIX_ARGS = -v 1eaf -p 0003
```
Comment (drashna) [inline keyboards/modelh/rules.mk:10]: Pretty sure that most (if not all) of this is set by specifying the bootloader and MCU. 
```suggestion
BOOTLOADER = stm32duino
```

File: keyboards/modelh/modelh.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+/* This is a shortcut to help you visually see your layout.
+ *
+ * The first section contains all of the arguments representing the physical
+ * layout of the board and position of the keys.
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+
+#define LAYOUT_fullsize_ansi_wkl( \
```
Comment (drashna) [inline keyboards/modelh/modelh.h:30]: these should be moved to info.json.

`qmk info -kb modelh -f json` will render a lot of that for you.

File: keyboards/modelh/chconf.h
Comment (drashna) [inline keyboards/modelh/chconf.h:1]: Not needed. Only values changed from default are needed.

File: keyboards/modelh/bootloader_defs.h
Comment (drashna) [inline keyboards/modelh/bootloader_defs.h:1]: not needed.

File: keyboards/modelh/config.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+
+/* Model H customizations */
+#define MODELH_LED_PINS { B9, B8, B7 }
```
Comment (drashna) [inline keyboards/modelh/config.h:22]:
```suggestion
```

File: keyboards/modelh/config.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+#define MODELH_STATUS_LED C13
+
+
+/* key matrix size */
+#define MATRIX_ROWS 8
+#define MATRIX_COLS 16
+
```
Comment (drashna) [inline keyboards/modelh/config.h:30]:
```suggestion
```

File: keyboards/modelh/halconf.h
Comment (drashna) [inline keyboards/modelh/halconf.h:1]: should be removed. only changed values are needed.

File: keyboards/modelh/info.json [hunk 5]
```diff
@@ -0,0 +1,1103 @@
...
+      "A1"
+    ]
+  },
+  "indicators": {
+    "num_lock": "B9",
+    "caps_lock": "B8",
+    "scroll_lock": "B7"
+  },
+  "processor": "STM32F103",
+  "bootloader": "stm32-dfu",
```
Comment (drashna) [inline keyboards/modelh/info.json:49]: this is at direct odds with the config in your rules.mk.

```suggestion
  "bootloader": "stm32duino",
```

File: keyboards/modelh/keymaps/default/keymap.c [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    KC_TAB,  KC_Q,    KC_W,   KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS,   KC_DEL,  KC_END,  KC_PGDN,   KC_P7,   KC_P8,   KC_P9,   KC_PPLS,
+    KC_CAPS, KC_A,    KC_S,   KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,                                 KC_P4,   KC_P5,   KC_P6,
+    KC_LSFT,          KC_Z,   KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH,          KC_RSFT,            KC_UP,              KC_P1,   KC_P2,   KC_P3,   KC_PENT,
+    KC_LCTL,          KC_LALT,                           KC_SPC,                                      KC_RALT,          KC_RCTL,   KC_LEFT, KC_DOWN, KC_RGHT,   KC_P0,            KC_PDOT
+  ),
+};
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+    return true;
+}
```
Comment (drashna) [inline keyboards/modelh/keymaps/default/keymap.c:31]:
```suggestion
```

File: keyboards/modelh/mcuconf.h
Comment (drashna) [inline keyboards/modelh/mcuconf.h:1]: same as other chibios config files

File: keyboards/modelh/modelh.c [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+void led_init_ports() {
+    setPinOutput(led_pins[0]);
+    setPinOutput(led_pins[1]);
+    setPinOutput(led_pins[2]);
+}
+
+#else
+void led_init_ports() { }
+#endif
```
Comment (drashna) [inline keyboards/modelh/modelh.c:40]:
```suggestion
```

File: keyboards/modelh/modelh.c [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+void keyboard_pre_init_kb(void) {
+    setPinOutput(MODELH_STATUS_LED);
+    writePin(MODELH_STATUS_LED, 0);
+
+    led_init_ports();
```
Comment (drashna) [inline keyboards/modelh/modelh.c:46]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by jberclaz** [comment]
Thanks for the comments. I addressed all of them (I think) and the code is much simpler now.

File: keyboards/modelh/config.h [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+//#define NO_PRINT
+
+/* disable action features */
+//#define NO_ACTION_LAYER
+//#define NO_ACTION_TAPPING
+//#define NO_ACTION_ONESHOT
+
+/* disable these deprecated features by default */
+//#define NO_ACTION_MACRO
+//#define NO_ACTION_FUNCTION
```
Comment (fauxpark) [inline keyboards/modelh/config.h:53]:
```suggestion
```

File: keyboards/modelh/info.json [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        "nkro": false,
+        "sleep_led": true
+    },
+    "indicators": {
+        "caps_lock": "B8",
+        "num_lock": "B9",
+        "scroll_lock": "B7"
+    },
+    "keyboard_folder": "modelh",
+    "keymaps": {},
```
Comment (fauxpark) [inline keyboards/modelh/info.json:27]:
```suggestion
```

File: keyboards/modelh/rules.mk [hunk 11]
```diff
@@ -0,0 +1,9 @@
+# MCU name
+MCU = STM32F103
+
+BOOTLOADER = stm32duino
+
```
Comment (fauxpark) [inline keyboards/modelh/rules.mk:5]:
```suggestion
```

File: keyboards/modelh/keymaps/default/readme.md [hunk 12]
```diff
@@ -0,0 +1 @@
+# The default keymap for modelh
```
Comment (fauxpark) [inline keyboards/modelh/keymaps/default/readme.md:1]: This file can be deleted.

File: keyboards/modelh/info.json [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+                { "label": "\u2193", "matrix": [7, 11], "w": 1, "x": 16.25, "y": 5.5 },
+                { "label": "\u2192", "matrix": [7, 12], "w": 1, "x": 17.25, "y": 5.5 },
+                { "label": "0", "matrix": [0, 12], "w": 2, "x": 18.5, "y": 5.5 },
+                { "label": ".", "matrix": [0, 13], "w": 1, "x": 20.5, "y": 5.5 }
+            ]
+        },
+        "LAYOUT_fullsize_iso_wkl": {
+            "c_macro": true,
+            "filename": "keyboards/modelh/modelh.h",
+            "json_layout": true,
```
Comment (fauxpark) [inline keyboards/modelh/info.json:166]:
```suggestion
```

File: keyboards/modelh/info.json [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+    "mouse_key": {
+        "enabled": false
+    },
+    "parse_errors": [],
+    "parse_warnings": [],
+    "platform": "STM32",
+    "platform_key": "chibios",
+    "processor": "STM32F103",
+    "processor_type": "arm",
+    "protocol": "ChibiOS",
```
Comment (fauxpark) [inline keyboards/modelh/info.json:46]:
```suggestion
    "processor": "STM32F103",
```

File: keyboards/modelh/info.json [hunk 15]
```diff
@@ -0,0 +1,273 @@
+{
+    "manufacturer": "IBM",
+    "keyboard_name": "Model M",
+    "maintainer": "jhawthorn",
+    "board": "STM32_F103_STM32DUINO",
+    "bootloader": "stm32duino",
+    "bootmagic": {
+        "enabled": false
+    },
+    "debounce": 5,
```
Comment (fauxpark) [inline keyboards/modelh/info.json:10]:
```suggestion
```

File: keyboards/modelh/config.h [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+
+
+/* Model H customizations */
+
+#define MODELH_STATUS_LED C13
+
+
+/* define if matrix has ghost (lacks anti-ghosting diodes) */
+#define MATRIX_HAS_GHOST
+
```
Comment (fauxpark) [inline keyboards/modelh/config.h:28]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by jberclaz** [comment]
Thanks for the review. Addressed.

File: keyboards/modelh/chconf.h [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+#define CH_CFG_USE_TM                       TRUE
+#define CH_CFG_USE_REGISTRY                 TRUE
+#define CH_CFG_USE_WAITEXIT                 TRUE
+#define CH_CFG_USE_CONDVARS                 TRUE
+#define CH_CFG_USE_CONDVARS_TIMEOUT         FALSE
+#define CH_CFG_USE_MESSAGES                 TRUE
+#define CH_CFG_USE_MAILBOXES                TRUE
+#define CH_CFG_USE_HEAP                     TRUE
+#define CH_CFG_USE_OBJ_CACHES               TRUE
+#define CH_CFG_USE_DELEGATES                TRUE
```
Comment (fauxpark) [inline keyboards/modelh/chconf.h:29]: Is there a particular reason for these?

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/modelh/info.json [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+        "sleep_led": true
+    },
+    "indicators": {
+        "caps_lock": "B8",
+        "num_lock": "B9",
+        "scroll_lock": "B7"
+    },
+    "matrix_pins": {
+        "cols": ["A10", "A9", "A8", "B15", "B14", "B13", "B12", "B11", "B10", "B1", "B0", "A7", "A6", "A5", "A4", "A3"],
+        "ghost": false,
```
Comment (fauxpark) [inline keyboards/modelh/info.json:23]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/modelh/chconf.h (hunk 17 above)
Comment (jberclaz) [reply keyboards/modelh/chconf.h:29]: I'm not the original author. The original code was made for an [old version ](https://github.com/jhawthorn/qmk_firmware/commits/modelh_stm32)of QMK. I just ported the changes here.
About this file specifically, I included the parameters that were different from the default ones in `platforms/chibios/boards/common/configs/chconf.h`.

**Comment by jberclaz** [comment]
I believe I have addressed all comments.

**Comment by jberclaz** [comment]
@fauxpark , anything else you want me to change?

**Comment by medoix** [comment]
@drashna is this PR able to be merged in yet please?

**Comment by hcbille** [comment]
Is there anything that could be done to get this moving?

File: keyboards/modelh/readme.md [hunk 19]
```diff
@@ -0,0 +1,19 @@
+# Model H - A USB capable PCB for the IBM Model M keyboard
+
+![modelh](https://modelh.club/modelh_assembled.jpg)
```
Comment (drashna) [inline keyboards/modelh/readme.md:3]: Could you rehost this on imgur or the like?

Comment (jberclaz) [reply keyboards/modelh/readme.md:3]: Sure, will do.

File: keyboards/modelh/chconf.h (hunk 17 above)
Comment (fauxpark) [reply keyboards/modelh/chconf.h:29]: Could you try removing this file and seeing if the board still works? I think most of these are pointless and inflate the file size for no reason; `CH_CFG_ST_TIMEDELTA` *may* be needed but usually isn't.

File: keyboards/modelh/mcuconf.h
Comment (fauxpark) [inline keyboards/modelh/mcuconf.h:1]: I don't see any features enabled which would require SPI or a timer. This file can likely be removed.

File: keyboards/modelh/info.json [hunk 20]
```diff
@@ -0,0 +1,283 @@
...
+    "bootloader": "stm32duino",
+    "diode_direction": "COL2ROW",
+    "features": {
+        "bootmagic": false,
+        "command": false,
+        "console": false,
+        "extrakey": true,
+        "mousekey": false,
+        "nkro": false,
+        "sleep_led": false
```
Comment (fauxpark) [inline keyboards/modelh/info.json:14]:
```suggestion
        "nkro": false
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/modelh/mcuconf.h
Comment (jberclaz) [reply keyboards/modelh/mcuconf.h:1]: When I remove it, I get a compilation error:
```
./lib/chibios/os/hal/ports/STM32/LLD/TIMv1/hal_pwm_lld.h:528:2: error: #error "PWM driver activated but no TIM peripheral assigned"
```

File: keyboards/modelh/rules.mk [hunk 21]
```diff
@@ -0,0 +1,2 @@
+# Enter lower-power sleep mode when on the ChibiOS idle thread
+OPT_DEFS += -DCORTEX_ENABLE_WFI_IDLE=TRUE
```
Comment (fauxpark) [inline keyboards/modelh/rules.mk:2]: This can be removed too, per https://github.com/qmk/qmk_firmware/pull/21353

File: keyboards/modelh/mcuconf.h
Comment (fauxpark) [reply keyboards/modelh/mcuconf.h:1]: Because PWM is enabled in halconf.h. That file can likely be removed as well.

Comment (jberclaz) [reply keyboards/modelh/mcuconf.h:1]: Ok, so remove mcuconf.h and the line `HAL_USE_PWM` in `halconf.h`? I'll give it a try.

Comment (jberclaz) [reply keyboards/modelh/mcuconf.h:1]: Ok. I just flashed the keyboard and it is still working without it. I'll remove it.

Comment (fauxpark) [reply keyboards/modelh/mcuconf.h:1]: As I've said, all three of these ChibiOS config files can probably be completely deleted.

Comment (jberclaz) [reply keyboards/modelh/mcuconf.h:1]: Sure, let me try that.

Comment (jberclaz) [reply keyboards/modelh/mcuconf.h:1]: Ok, looks also fine without.

**Review by fauxpark — APPROVED** [review]

**Comment by hcbille** [comment]
@drashna , what is needed from you before you can approve and this can be merged?

**Review by drashna — APPROVED** [review]

**Comment by jberclaz** [comment]
Thanks for the reviews!

**Comment by sigprof** [comment]
Should this board have been placed under `ibm/model_m/`, as per #14996 ?

**Comment by jberclaz** [comment]
> Should this board have been placed under `ibm/model_m/`, as per #14996 ?

Oh, right. I'll make another PR.

**Pull request merged**